### PR TITLE
fixed expander bug, builtin optimisation

### PR DESCRIPTION
**Pull request by robertrinh**
1. fixed issue where echo $USER no quotes would segfault
2. updated export regarding checks
3. exit builtin optimalised
4. minor error_message fixes
5. minor free's
6. sources/executor/execute_commands/execute_piped_command.c check for main builtin needs to be deleted (forgot :))

### CODE REVIEW

File: sources/builtins/export.c [hunk 1]
```diff
@@ -128,10 +130,14 @@ int		export(t_cmd *cmd, t_shell *shell)
 		return (show_error_message(E_EXPORT, C_RED, "", X_EXPORT));
 	while (i < cmd->arg_count)
 	{
-		if (is_valid_export_arg(cmd->args[i]))
-			add_arg_to_env(shell, cmd->args[i]);
-		else
-			show_error_message(E_EXPORT, C_RED, cmd->args[i], X_EXPORT);
+			if (is_valid_export_arg(cmd->args[i]))
```
Comment (QBeukelman) [inline sources/builtins/export.c:133]: `is_valid_export_arg()` should handle all checks. `count_delimiters()` should not be here.

File: sources/executor/execute_commands/execute_piped_command.c [hunk 2]
```diff
@@ -22,9 +22,13 @@ t_validation	execute_piped_command(t_shell *shell, t_cmd *cmd)
 	cmd_path = cmd->cmd_path;
 	formatted_cmd = cmd->formatted_cmd;
 
+	if (is_builtin(shell->builtin_main, cmd, B_NUM_MAIN))
```
Comment (QBeukelman) [inline sources/executor/execute_commands/execute_piped_command.c:25]: Are these commands not meant to execute in the main thread?

File: sources/executor/execute_commands/execute_wait.c [hunk 3]
```diff
@@ -6,25 +6,17 @@
...
 /*                                                   +#+                      */
 /*   Created: 2024/04/22 21:17:34 by quentinbeuk   #+#    #+#                 */
-/*   Updated: 2024/04/23 21:05:11 by quentinbeuk   ########   odam.nl         */
+/*   Updated: 2024/04/24 19:19:49 by qtrinh        ########   odam.nl         */
 /*                                                                            */
 /* ************************************************************************** */
 
 #include "../../../includes/minishell.h"
 
-static int	handle_signal(int signal)
```
Comment (QBeukelman) [inline sources/executor/execute_commands/execute_wait.c:15]: Keep this function. I am still working on the exit codes here.

File: sources/executor/execute_commands/execute_wait.c [hunk 4]
```diff
@@ -41,12 +33,10 @@ int	wait_for_last_cmd(int child_count, int last_pid)
 			show_error_message(E_WAITPID, C_RED, "", X_WAITPID);
 		if (waitpid_ret == last_pid)
 			last_status = status;
-		if (WIFEXITED(status) || WIFSIGNALED(status))
+		if (WIFEXITED(status))
 			i++;
 	}
 	if (WIFEXITED(last_status))
 		return (WEXITSTATUS(last_status));
-	else if (WIFSIGNALED(last_status))
```
Comment (QBeukelman) [inline sources/executor/execute_commands/execute_wait.c:49]: Keep this

File: sources/expander/expander.c [hunk 5]
```diff
@@ -37,8 +37,9 @@ static char	*expand_arg(char **env, char *arg, size_t i)
 		env_utils->key = ft_strjoin("$", env_utils->key);
 	if (arg && env_utils->value && is_arg_key(arg, env_utils->key))
 	{
-		free_env_values(env_utils->key, env_utils->value, env_utils);
-		return (arg = env_utils->value);
+		arg = env_utils->value;
```
Comment (QBeukelman) [inline sources/expander/expander.c:40]: All values need to be passed to`free_env_values()` function.

**Review by QBeukelman — CHANGES_REQUESTED** [review]
* Incorrect branch name: "bugfix/#102-Branch-name"
* Incorrect commit message: "bugfix/#102 Commit message"

File: sources/executor/execute_commands/execute_piped_command.c (hunk 2 above)
Comment (robertrinh) [reply sources/executor/execute_commands/execute_piped_command.c:25]: Correct, I have mentioned that in the comment -> bulletpoint 6

File: sources/executor/execute_commands/execute_wait.c (hunk 3 above)
Comment (robertrinh) [reply sources/executor/execute_commands/execute_wait.c:15]: my bad! My opinion is that the signals are handled correctly in the childprocess and I dont believe it's neccesary to handle the signals at waitpid function. will revert it

File: sources/executor/execute_commands/execute_wait.c (hunk 4 above)
Comment (robertrinh) [reply sources/executor/execute_commands/execute_wait.c:49]: see previous comment

File: sources/expander/expander.c (hunk 5 above)
Comment (robertrinh) [reply sources/expander/expander.c:40]: Issue is that freeing the values in this specific statement that the arg value would be lost, having no value.

[minishell]: echo $USER
� 
[minishell]: echo "$USER"
qtrinh 

you can try it out. not passing the env_utils->value to free_env_values retains the expanded env value and therefore, works correctly.

File: sources/builtins/export.c (hunk 1 above)
Comment (robertrinh) [reply sources/builtins/export.c:133]: fair point. Issue was that export without the `export_delimiter` would segfault. Bash behavior would be this:

export a -> does nothing
export a in our minishell -> segfaults.

The idea was that the function should `return(0)` when no delimiter is found (and not an error!), essentially not doing anything like bash. Maybe you know a better approach? :O

Comment (robertrinh) [reply sources/builtins/export.c:133]: edit: fixed

Comment (robertrinh) [reply sources/builtins/export.c:133]: On second note: I put the only characters check in `is_valid_export_key`. As bash doesn't necessarily do anything with it, we can just argue that we handle that as an error instead. holds up the same result, not exporting anything in the env!

**Comment by robertrinh** [comment]
1. The stuff in the prior comments are updated accordingly except expander (will discuss!)
2. fixed issue #100 where export stops giving error after multiple inputs
3. fixed issue mentioned in #85 that the minishell outputted previous command when pressing `SPACEBAR`
4. Started combatting memory issues, values initialised
5. Fixed few memory issues consisting `''Conditional jump or move depends on uninitialised value(s)''`
6. fixed issue where original_fd kept incrementing after a command
7. #85 in progress.
8. removed fd_err in the cmd struct, as we have not used it at all.

**Pull request merged**